### PR TITLE
add palette detection to fjxl

### PR DESCRIPTION
**Pull request by jonsneyers**
Makes it a bit slower in general, though not measurably in case of photo (which will have a palette collision early).

Significantly better density in case palette can actually be used.

### CODE REVIEW

File: experimental/fast_lossless/fast_lossless.cc [hunk 1]
```diff
@@ -872,6 +878,71 @@ void WriteACSection(const unsigned char* rgba, size_t x0, size_t y0, size_t oxs,
...
+  encoder.output = &output;
+  encoder.code = &code;
+
+  constexpr size_t kPadding = 16;
+
+  int16_t group_data[2][256 + kPadding * 2] = {};
+
+  for (size_t y = 0; y < ys; y++) {
+    // Pre-fill rows with palette converted pixels.
+    const uint32_t* p = reinterpret_cast<const uint32_t*>(
```
Comment (veluca93) [inline experimental/fast_lossless/fast_lossless.cc:918]: let's avoid reinterpret_casts

File: experimental/fast_lossless/fast_lossless.cc [hunk 2]
```diff
@@ -886,17 +957,156 @@ void CollectSamples(const unsigned char* rgba, size_t x0, size_t y0,
...
+  output->Write(2, 0b01);     // Palette
+  output->Write(5, 0b00000);  // Starting from ch 0
+  output->Write(2, 0b10);     // 4-channel palette (RGBA)
+  size_t pcolors = (pcolors_real + kChunkSize - 1) / kChunkSize * kChunkSize;
+  // pcolors > 0 and pcolors <= 1024
+  if (pcolors < 256) {
+    output->Write(2, 0b00);
+    output->Write(8, pcolors);
+  } else {
+    output->Write(2, 0b01);
```
Comment (veluca93) [inline experimental/fast_lossless/fast_lossless.cc:976]: static_assert here that kMaxColors is <= 1024?

File: experimental/fast_lossless/fast_lossless.cc [hunk 3]
```diff
@@ -886,17 +957,156 @@ void CollectSamples(const unsigned char* rgba, size_t x0, size_t y0,
...
+    if (palette[0] == 1) palette[0] = 0;
+    std::sort(palette.begin(), palette.end(), [](uint32_t a, uint32_t b) {
+      if (a == 0) return false;
+      if (b == 0) return true;
+      uint8_t a_rgba[4];
+      uint8_t b_rgba[4];
+      memcpy(a_rgba, &a, 4);
+      memcpy(b_rgba, &b, 4);
+      float a_y =
+          (0.299 * a_rgba[0] + 0.587 * a_rgba[1] + 0.114 * a_rgba[2] + 0.01) *
```
Comment (veluca93) [inline experimental/fast_lossless/fast_lossless.cc:1080]: probably want to do operations in floats (0.299f)

File: experimental/fast_lossless/fast_lossless.cc [hunk 4]
```diff
@@ -886,17 +957,156 @@ void CollectSamples(const unsigned char* rgba, size_t x0, size_t y0,
...
+    bool have_color = false;
+    uint8_t minG = 255, maxG = 0;
+    for (; k < kMaxColors; k++) {
+      if (palette[k] == 0) break;
+      uint8_t* rgba = reinterpret_cast<uint8_t*>(&palette[k]);
+      if (rgba[0] != rgba[1] || rgba[0] != rgba[2]) have_color = true;
+      lookup[pixel_hash(palette[k])] = pcolors++;
+      if (rgba[1] < minG) minG = rgba[1];
+      if (rgba[1] > maxG) maxG = rgba[1];
+      // printf("Palette entry %i : %08X\n",k,palette[k]);
```
Comment (veluca93) [inline experimental/fast_lossless/fast_lossless.cc:1096]: stale comment

File: experimental/fast_lossless/fast_lossless.cc (hunk 1 above)
Comment (jonsneyers) [reply experimental/fast_lossless/fast_lossless.cc]: done

File: experimental/fast_lossless/fast_lossless.cc (hunk 2 above)
Comment (jonsneyers) [reply experimental/fast_lossless/fast_lossless.cc:998]: done

File: experimental/fast_lossless/fast_lossless.cc (hunk 3 above)
Comment (jonsneyers) [reply experimental/fast_lossless/fast_lossless.cc]: indeed.

File: experimental/fast_lossless/fast_lossless.cc (hunk 4 above)
Comment (jonsneyers) [reply experimental/fast_lossless/fast_lossless.cc]: removed

**Comment by jonsneyers** [comment]
Made some further changes, PTAL.

This adds an `effort` parameter (default is 2), which controls the amount of sampling done (and also to disable palette, if effort is set to 0 or 1). The number of pixels that are sampled is something close to `total_pixels * effort / 128`.

**Review by veluca93 — APPROVED** [review]

**Pull request merged**